vmdb/app/models/ems_refresh/parsers/openstack_infra.rb

```diff
@@ -0,0 +1,257 @@
+module EmsRefresh
+  module Parsers
+    class OpenstackInfra < Infra
+      def self.ems_inv_to_hashes(ems, options = nil)
+        new(ems, options).ems_inv_to_hashes
+      end
+
+      def initialize(ems, _options = nil)
+        @ems               = ems
+        @connection        = ems.connect
+        @data              = {}
+        @data_index        = {}
+        @host_hash_by_name = {}
+
+        @known_flavors = Set.new
+
+        @os_handle                  = ems.openstack_handle
+        @compute_service            = @connection # for consistency
+        @baremetal_service          = @os_handle.detect_baremetal_service
+        @baremetal_service_name     = @os_handle.baremetal_service_name
+        @orchestration_service      = @os_handle.detect_orchestration_service
+        @orchestration_service_name = @os_handle.orchestration_service_name
+      end
+
+      def ems_inv_to_hashes
+        log_header = "MIQ(#{self.class.name}.#{__method__}) Collecting data" \
+                     " for EMS name: [#{@ems.name}] id: [#{@ems.id}]"
+        $fog_log.info("#{log_header}...")
+
+        load_hosts
+        load_orchestration_stacks
+
+        $fog_log.info("#{log_header}...Complete")
+        @data
+      end
+
+      private
+
+      def servers
+        @servers ||= @connection.servers_for_accessible_tenants
+      end
+
+      def hosts
+        @hosts ||= @baremetal_service.nodes.details
+      end
+
+      def hosts_ports
+        @hosts_ports ||= @baremetal_service.ports.details
+      end
+
+      def stacks
+        @stacks ||= detailed_stacks
+      end
+
+      def detailed_stacks
+        detailed_stacks = []
+        # Seems like Heat doesn't suport detailed list, so we need to call get for each stack, which is suboptimal
+        @orchestration_service.stacks.each { |stack| detailed_stacks << stack.details }
+        detailed_stacks
+      end
+
+      def resources
+        return @resources if @resources
+
+        resources = []
+        stacks.each do |stack|
+          # Nested depth 50 just for sure, although nobody should nest templates that much
+          all_stack_resources = @orchestration_service.list_resources(stack, :nested_depth => 50)
```

Comment:
Yeah this parameter will be needed, otherwise we would have to do a lot of queries to get the nested resources.

Btw. would be nice if you could explain me the concept of nested stacks in AWS. Heat has also nested stacks implemented using resources. You can get the flat list with ^ nested depth, or you can do heat resource-list stack_name resource_id. When it is nested stack the the stack name is parent resource or stack and resource id is the nested stack you want to list. Now you can pass only stack object to resource list in Fog, which is not correct for nested queries.